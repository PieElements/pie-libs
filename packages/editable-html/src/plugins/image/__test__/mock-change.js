import { Data } from 'slate';
import { stub } from 'sinon';

export default function MockChange() {

  this.setNodeByKey = stub().returns(this);

  this.removeNodeByKey = stub().returns(this);
  this.insertInline = stub().returns(this);
}

export function MockDocument() {

  this.getChild = stub().returns({ data: Data.create({}) });

  this.getDescendant = stub();
}