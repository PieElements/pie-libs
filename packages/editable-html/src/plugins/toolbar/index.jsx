import EditorAndToolbar from './editor-and-toolbar';
import React from 'react';

/**
 * Adds a toolbar which is shown when the editor is focused. 
 * @param {*} opts 
 */
export default function ToolbarPlugin(opts) {
  return {
    renderEditor: props => <EditorAndToolbar
      {...props}
      onDone={opts.onDone} />
  }
}

