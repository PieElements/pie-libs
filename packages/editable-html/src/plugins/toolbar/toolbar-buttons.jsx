import React from 'react';
import debug from 'debug';
import injectSheet from 'react-jss';

const styles = {
  button: {
    color: 'grey',
    display: 'inline-flex',
    padding: '2px',
    '& :hover': {
      color: 'black'
    }
  },
  active: {
    color: 'black'
  }
}

const buttonStyles = {
  root: Object.assign(styles.button, {
    display: 'inline-flex'
  })
}

const rbLog = debug('editable-html:raw-button');

export class RawButton extends React.Component {

  constructor(props) {
    super(props)

    this.onClick = (e) => {
      rbLog('[onClick]');
      e.preventDefault();
      const { onClick } = this.props;
      onClick(e)
    }
  }

  render() {
    const { classes, children } = this.props;
    return <div
      onMouseDown={this.onClick}
      className={classes.root}>{children}</div>;
  }
}

export const Button = injectSheet(buttonStyles)(RawButton);


export class RawMarkButton extends React.Component {
  constructor(props) {
    super(props);

    this.onToggle = (e) => {
      e.preventDefault();
      this.props.onToggle(this.props.mark);
    };
  }

  render() {
    const { classes, children, active } = this.props;
    let className = classes.button;

    if (active) {
      className += ` ${classes.active}`;
    }

    return <span
      className={className}
      onMouseDown={this.onToggle}>{children}</span>
  }
}

export const MarkButton = injectSheet(styles)(RawMarkButton);