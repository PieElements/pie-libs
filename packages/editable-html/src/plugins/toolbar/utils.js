export const findSingleNode = (value) => {

  if (!value || !value.isCollapsed || !value.startKey) {
    return;
  }

  const inline = value.document.getClosestInline(value.startKey);

  if (inline) {
    return inline;
  }

  const block = value.document.getClosestBlock(value.startKey);

  if (block) {
    return block;
  }

}