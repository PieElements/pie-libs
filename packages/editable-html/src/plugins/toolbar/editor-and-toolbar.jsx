import { findSingleNode, selectedNode } from './utils';

import React from 'react';
import Toolbar from './toolbar';
import classNames from 'classnames';
import debug from 'debug';
import injectSheet from 'react-jss';
import { primary } from '../../theme';

const log = debug('editable-html:plugins:toolbar:editor-and-toolbar');

export class RawEditorAndToolbar extends React.Component {

  render() {

    const {
     classes,
      children,
      value,
      plugins,
      onChange,
      onDone,
      focusedNode
   } = this.props;


    const inFocus = value.isFocused || (focusedNode !== null && focusedNode !== undefined);
    const holderNames = classNames(classes.editorHolder, inFocus && classes.editorInFocus);

    log('[render] inFocus: ', inFocus, 'value.isFocused:', value.isFocused, 'focused node: ', focusedNode);

    return (
      <div className={classes.root}>
        <div className={holderNames}>{children}</div>
        <Toolbar
          plugins={plugins}
          value={value}
          isFocused={inFocus}
          onChange={onChange}
          onDone={onDone} />
      </div>
    );
  }
}

const style = {
  root: {
    position: 'relative',
    padding: '0px',
    border: 'none',
    borderBottom: '0px solid #cccccc',
    borderRadius: '0px',
    cursor: 'text',
    '& [data-slate-editor="true"]': {
      overflow: 'auto',
      maxHeight: '500px',
    }
  },
  editorHolder: {
    position: 'relative',
    padding: '7px',
    '&::before': {
      left: '0',
      right: '0',
      bottom: '0',
      height: '1px',
      content: '""',
      position: 'absolute',
      transition: 'background-color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      pointerEvents: 'none',
      backgroundColor: 'rgba(0, 0, 0, 0.42)'
    },
    '&::after': {
      left: '0',
      right: '0',
      bottom: '0',
      height: '1px',
      content: '""',
      position: 'absolute',
      transform: 'scaleX(0)',
      transition: `transform 200ms cubic-bezier(0.0, 0.0, 0.2, 1) 0ms, background-color 200ms linear`,
      backgroundColor: 'rgba(0, 0, 0, 0.42)',
    },
    '&:focus': {
      '&::after': {
        transform: 'scaleX(1)',
        backgroundColor: primary,
        height: '2px'
      }
    },
    '&:hover': {
      '&::after': {
        transform: 'scaleX(1)',
        backgroundColor: 'black',
        height: '2px'
      }
    },
  },
  editorInFocus: {
    '&::after': {
      transform: 'scaleX(1)',
      backgroundColor: primary,
      height: '2px'
    },
    '&:hover': {
      '&::after': {
        backgroundColor: primary
      }
    }
  }
}

export default injectSheet(style)(RawEditorAndToolbar);
