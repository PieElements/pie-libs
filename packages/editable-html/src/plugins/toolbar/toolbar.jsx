import { Button, MarkButton } from './toolbar-buttons';

import Bold from 'material-ui-icons/FormatBold';
import Check from 'material-ui-icons/Check';
import Code from 'material-ui-icons/Code';
import Delete from 'material-ui-icons/Delete';
import Functions from 'material-ui-icons/Functions';
import IconButton from 'material-ui/IconButton';
import Image from 'material-ui-icons/Image';
import Italic from 'material-ui-icons/FormatItalic';
import PropTypes from 'prop-types';
import React from 'react';
import Strikethrough from 'material-ui-icons/FormatStrikethrough';
import Underlined from 'material-ui-icons/FormatUnderlined';
import _ from 'lodash';
import classNames from 'classnames';
import debug from 'debug';
import { findSingleNode } from './utils';
//TODO: use mui createStyleSheet (ensures the class overrides work).
import injectSheet from 'react-jss';

const log = debug('editable-html:plugins:toolbar');

const toolbarStyle = {
  toolbar: {
    position: 'absolute',
    zIndex: 10,
    cursor: 'pointer',
    justifyContent: 'space-between',
    background: 'var(--editable-html-toolbar-bg, #efefef)',
    margin: '0px',
    padding: '2px',
    width: '100%',
    boxShadow: '0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)',
    boxSizing: 'border-box',
    display: 'none'
  },
  focused: {
    display: 'flex',
  },
  iconRoot: {
    width: '28px',
    height: '28px'
  },
  label: {
    color: 'var(--editable-html-toolbar-check, #00bb00)'
  }
}

const ToolbarButton = (props) => {

  const hasMark = (type) => {
    const { value } = props;
    return value.marks.some(mark => mark.type == type)
  }

  if (props.isMark) {
    const isActive = hasMark(props.type);
    return <MarkButton
      active={isActive}
      label={props.type}
      onToggle={() => {
        const c = props.onToggle(props.value.change())
        props.onChange(c);
      }}
      mark={props.type}
    >{props.icon}</MarkButton>
  } else {
    return <Button
      onClick={() => props.onClick(props.value, props.onChange)}>{props.icon}</Button>
  }
}

const RawDefaultToolbar = ({ plugins, value, onChange, classes }) => {
  const toolbarPlugins = plugins.filter(p => p.toolbar).map(p => p.toolbar);
  return (
    <div className={classes.inline}>
      {toolbarPlugins.map((p, index) => {
        return <ToolbarButton
          {...p}
          key={index}
          value={value}
          onChange={onChange} />
      })}
    </div>);
}

const DefaultToolbar = injectSheet(toolbarStyle)(RawDefaultToolbar);

export class RawToolbar extends React.Component {

  constructor(props) {
    super(props);
    this.hasMark = (type) => {
      const { value } = this.props;
      return value.marks.some(mark => mark.type == type)
    }

    this.hasBlock = (type) => {
      const { value } = this.props;
      return value.blocks.some(node => node.type == type)
    }
  }

  onToggle = (plugin) => {
    const { value, onChange } = this.props;

    if (!plugin.onToggle) return;

    const change = plugin.onToggle(value.change())
    onChange(change);
  }

  onClick = (e) => {
    log('[onClick]')
    e.preventDefault();
  }

  onButtonClick = (fn) => {
    return e => {
      e.preventDefault();
      fn();
    }
  }

  render() {
    const {
      classes,
      onDone,
      zIndex,
      onFocus,
      onBlur,
      plugins,
      value,
      onChange,
      isFocused,
      onClick } = this.props;

    const node = findSingleNode(value);

    log('[render] node: ', node);

    const plugin = plugins.find(p => {
      if (!node) {
        return;
      }

      if (p.toolbar) {
        return p.toolbar.supports && p.toolbar.supports(node);
      }
    });
    log('[render] plugin: ', plugin);

    const CustomToolbar = plugin && plugin.toolbar && plugin.toolbar.customToolbar ? plugin.toolbar.customToolbar(node) : null;

    log('[render] CustomToolbar: ', CustomToolbar);

    const style = zIndex ? { zIndex } : {};

    const names = classNames(classes.toolbar, isFocused && classes.focused)

    const doneFn = this.onButtonClick(onDone);

    const deletable = node && plugin && plugin.deleteNode;

    return (
      <div className={names}
        onClick={this.onClick}>
        {CustomToolbar ?
          <CustomToolbar
            value={value}
            onChange={onChange}
            node={node} /> :
          <DefaultToolbar
            plugins={plugins}
            value={value}
            onChange={onChange} />}

        <div className={classes.shared} >
          {deletable && <IconButton
            aria-label="Delete"
            style={{ width: '28px', height: '28px' }}
            className={classes.iconRoot}
            onClick={(e) => plugin.deleteNode(e, node, value, onChange)}
            classes={{
              root: classes.iconRoot
            }}>
            <Delete />
          </IconButton>}
          <IconButton
            aria-label="Done"
            style={{ width: '28px', height: '28px' }}
            className={classes.iconRoot}
            onClick={(plugin && plugin.onDone) ? (e) => plugin.onDone(e, node, value, onChange, doneFn) : doneFn}
            classes={{
              label: classes.label,
              root: classes.iconRoot
            }}>
            <Check />
          </IconButton>
        </div>
      </div >
    );
  }
}

RawToolbar.propTypes = {
  zIndex: PropTypes.number,
  value: PropTypes.object.isRequired,
  plugins: PropTypes.array,
  onImageClick: PropTypes.func,
  onDone: PropTypes.func.isRequired
}

export default injectSheet(_.clone(toolbarStyle), { index: 1000 })(RawToolbar);