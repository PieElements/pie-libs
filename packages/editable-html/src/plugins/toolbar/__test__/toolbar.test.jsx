import { classObject, mockIconButton, mockMathInput } from '../../../__test__/utils';

import { Data } from 'slate';
import { RawToolbar } from '../toolbar';
import React from 'react';
import debug from 'debug';
import renderer from 'react-test-renderer';
import { stub } from 'sinon';

mockMathInput();

jest.mock('material-ui/IconButton', () => {
  return (props) => <div className={props.className} style={props.style} ariaLabel={props['aria-label']}></div>
});

const log = debug('editable-html:test:toolbar');


describe('toolbar', () => {

  let onDelete, classes;

  beforeEach(() => {
    onDelete = stub();
    classes = classObject('iconRoot', 'inline', 'toolbar', 'focused', 'shared', 'inline')
  });

  test('renders custom toolbar', () => {

    const node = {}

    const value = {
      isCollapsed: true,
      startKey: 1,
      document: {
        getClosestInline: stub().returns(node)
      }
    }

    const plugins = [
      {
        deleteNode: () => true,
        toolbar: {
          supports: () => true,
          customToolbar: () => () => <div> --------- custom toolbar ----------- </div>
        }
      }
    ];

    const tree = renderer
      .create(<RawToolbar
        plugins={plugins}
        classes={classes}
        value={value}
        onDone={() => ({})} />).toJSON();

    log('tree: ', JSON.stringify(tree, null, '  '));
    expect(tree).toMatchSnapshot();
  });

  describe('default', () => {
    let plugins;

    beforeEach(() => {
      plugins = [
        {
          toolbar: {}
        }
      ];
    });

    test('renders default toolbar', () => {

      const tree = renderer
        .create(<RawToolbar
          plugins={plugins}
          classes={classes}
          value={{}}
          onDone={() => ({})} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

  });
});
