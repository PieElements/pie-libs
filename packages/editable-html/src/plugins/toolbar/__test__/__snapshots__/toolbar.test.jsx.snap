// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toolbar default renders default toolbar 1`] = `
<div
  className="toolbar"
  onClick={[Function]}
>
  <div
    className={undefined}
  >
    <div
      className="root-0-4"
      onMouseDown={[Function]}
    />
  </div>
  <div
    className="shared"
  >
    <div
      ariaLabel="Done"
      className="iconRoot"
      style={
        Object {
          "height": "28px",
          "width": "28px",
        }
      }
    />
  </div>
</div>
`;

exports[`toolbar renders custom toolbar 1`] = `
<div
  className="toolbar"
  onClick={[Function]}
>
  <div>
     --------- custom toolbar ----------- 
  </div>
  <div
    className="shared"
  >
    <div
      ariaLabel="Delete"
      className="iconRoot"
      style={
        Object {
          "height": "28px",
          "width": "28px",
        }
      }
    />
    <div
      ariaLabel="Done"
      className="iconRoot"
      style={
        Object {
          "height": "28px",
          "width": "28px",
        }
      }
    />
  </div>
</div>
`;
